arwen/elrondapi/elrondei.go

```diff
@@ -730,12 +730,13 @@ func transferESDTExecute(
 		contractCallInput.ESDTTokenName = tokenIdentifier
 	}
 
-	err = output.TransferESDT(dest, sender, tokenIdentifier, big.NewInt(0).SetBytes(valueBytes), contractCallInput, uint64(gasLimit))
+	gasLimitForExec, err := output.TransferESDT(dest, sender, tokenIdentifier, big.NewInt(0).SetBytes(valueBytes), contractCallInput)
 	if arwen.WithFault(err, context, runtime.ElrondAPIErrorShouldFailExecution()) {
 		return 1
 	}
 
 	if host.AreInSameShard(sender, dest) && contractCallInput != nil && host.Blockchain().IsSmartContract(dest) {
+		contractCallInput.GasProvided = gasLimitForExec
 		_, _, _, err = host.ExecuteOnDestContext(contractCallInput)
 		if arwen.WithFault(err, context, runtime.ElrondSyncExecAPIErrorShouldFailExecution()) {
 			return 1
```

Comment:
we need to revert the ESDT transfer here in case of error. without using actual gas. there is already a function in host which does this.

Reply:
Added call to `host.RevertESDTTransfer()` after exporting it.